rust/src/lib.rs

```diff
@@ -289,52 +312,108 @@ pub unsafe extern "C" fn zcashlc_derive_extended_spending_keys(
 
 #[repr(C)]
 pub struct FFIUnifiedViewingKey {
+    account_id: u32,
     extfvk: *const c_char,
     extpub: *const c_char,
 }
 
+// TODO: Return these on init_accounts_table instead of just Sapling SKs
+// #[repr(C)]
+// pub struct FFIUnifiedSpendingKey {
+//     account_id: u32,
+//     transparent: *const u8,
+//     sapling: *const u8,
+// }
+
+
+// #[repr(C)]
+// pub struct FFIUSKBoxedSlice {
+//     ptr: *mut FFIUnifiedSpendingKey,
+//     len: usize, //number of elements
+// }
+
 #[repr(C)]
 pub struct FFIUVKBoxedSlice {
     ptr: *mut FFIUnifiedViewingKey,
     len: usize, // number of elems
 }
 
+fn ufvk_to_ffi(network: &Network, ufvk: &UnifiedFullViewingKey) -> FFIUnifiedViewingKey {
+    let encoded_extfvk = ufvk.sapling().map(|extfvk| {
+        CString::new(encode_extended_full_viewing_key(
+            network.hrp_sapling_extended_full_viewing_key(),
+            extfvk,
+        ))
+        .unwrap()
+        .into_raw()
+    });
+    let encoded_pubkey = ufvk.transparent().map(|t| {
+        let tkey_bytes: Vec<u8> = t.serialize();
+        CString::new(hex::encode(tkey_bytes)).unwrap().into_raw()
+    });
 
-fn unified_viewing_key_new(
-    extfvk: &ExtendedFullViewingKey, 
-    extpub: &PublicKey,
-    network: Network) -> FFIUnifiedViewingKey {
-    
-    let encoded_extfvk = encode_extended_full_viewing_key(
-        network.hrp_sapling_extended_full_viewing_key(),
-        extfvk,
-    );
-    let encoded_pubkey = hex::encode(&extpub.serialize()); 
-    
     FFIUnifiedViewingKey {
-        extfvk: CString::new(encoded_extfvk).unwrap().into_raw(),
-        extpub: CString::new(encoded_pubkey).unwrap().into_raw()
+        account_id: u32::from(ufvk.account()),
+        extfvk: encoded_extfvk.unwrap_or(std::ptr::null_mut()),
+        extpub: encoded_pubkey.unwrap_or(std::ptr::null_mut()),
     }
 }
+use std::convert::TryInto;
+
+fn ufvk_from_ffi(
+    network: &Network,
+    ffiufvk: &FFIUnifiedViewingKey,
+) -> Option<UnifiedFullViewingKey> {
+    //TODO
```

Comment:
What is this TODO about? Please file an issue and link to it, or remove the TODO.

Reply:
Thanks for pointing this out!